app/models/web_hook_events_daily_aggregate.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class WebHookEventsDailyAggregate < ActiveRecord::Base
+  belongs_to :web_hook
+
+  default_scope { order("created_at DESC") }
+  after_create :aggregate
+
+  def self.purge_old
+    where("created_at < ?", SiteSetting.retain_web_hook_events_period_days.days.ago).delete_all
+  end
+
+  def self.by_day(start_date, end_date, web_hook_id = nil)
+    result = where("date >= ? AND date <= ?", start_date.to_date, end_date.to_date)
+    result = result.where(web_hook_id: web_hook_id) if web_hook_id
+    result
+  end
+
+  def aggregate
+    events =
+      WebHookEvent.where(
+        "created_at >= ? AND created_at < ? AND web_hook_id = ?",
+        self.date,
+        self.date + 1.day,
+        self.web_hook_id,
+      )
+
+    self.mean_duration = events.sum(:duration) / events.count if events.count > 0
+
+    self.successful_events_id = events.where("status >= 200 AND status <= 299").pluck(:id)
+    self.failed_events_id = events.where("status < 200 OR status > 299").pluck(:id)
```

Comment:
While I appreciate your ideas for adding the extra precision here, it's just not necessary for an aggregate table. These `WebHookEvent` records are deleted daily based on the `retain_web_hook_events_period_days` site setting, so these IDs will end up pointing to nothing.

In the period before `retain_web_hook_events_period_days` ends up deleting the events, people who want a more granular view of things should look in the webhook event logs.

Reply:
However re-examining this and the spec I wrote on Meta, I actually think 1 record per day per webhook will be fine rather than 2, you could store `successful_event_count` and `failed_event_count` on this record.